### PR TITLE
Add ClassAccount storage to unique pallet

### LINKED ISSUE #9777: feature request: Storage query of uniques classes by account
We are working on UI for minting uniques assets and noticed that there's no way to query Classes by account, only assets.  I can query all asset keys first by account, then query all classes from there and filter by those where issuer === accountId, but this won't work for Classes that have no assets (empty classes)

So I am wondering if it possible to add such storage query to substrate.

The reason why it's a problem. Imagine a UI/UX where you first create a collection (class) and then mint NFTs (assets) into this collection.

Imagine if user refreshes a page between signing class create tx and asset mint tx, so now there's no way to display all his previously created empty classes to select the one where to mint first asset into

### PR DESCRIPTION
**Pull request by hamidra**
The changes addresses #9777 by adding a storage to unique pallet to keep classes that are owned by an account. The current Account storage keeps the instances that are owned by an account but currently the only way to query the classes that an account owns is by iterating through all asset classes. Having the account classes indexed seems to be very useful for NFT UXs which the user creates a collection of NFTs, and then needs to make changes to those collections in future.
[fixes #9777 ]

### CODE REVIEW

**Comment by xlc** [comment]
this requires migration

**Comment by hamidra** [comment]
Try-runtime was ran against a substrate node with some unique classes and assets and the migration result was complete with no errors:


2021-10-31 15:52:22 ✅ no migration for TransactionStorage    
2021-10-31 15:52:22 ⚠️ Uniques declares internal migrations (which *might* execute). On-chain `<wasm:stripped>` vs current storage version `<wasm:stripped>`    
2021-10-31 15:52:22 Running migration storage v1 for uniques with storage version <wasm:stripped>    
2021-10-31 15:52:22 Running migration storage v1 for uniques with storage version <wasm:stripped> was complete    
2021-10-31 15:52:22 ✅ no migration for Gilt

File: frame/uniques/src/migration.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline frame/uniques/src/migration.rs:3]:
```suggestion
// Copyright (C) 2021 Parity Technologies (UK) Ltd.
```

File: frame/uniques/src/migration.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	log::info!(
+		target: "runtime::uniques",
+		"Running migration storage v1 for uniques with storage version {:?}",
+		on_chain_storage_version,
+	);
+
+	if on_chain_storage_version < 1 {
+		let mut count = 0;
+		for (class, detail) in Class::<T, I>::iter() {
+			ClassAccount::<T, I>::insert(&detail.owner, &class, ());
```
Comment (kianenigma) [inline frame/uniques/src/migration.rs:37]: Note that this migration might be very big and cause issues on a parachain.

**Review by kianenigma — COMMENTED** [review]
Looks overall good to me. To test the migration properly with try-runtime you need to integrate it in cumulus, but I think the logic of the migration is pretty trivial. I am just worried about the size of migration.

**Comment by kianenigma** [comment]
> Try-runtime was ran against a substrate node with some unique classes and assets and the migration result was complete with no errors:
> 
> 2021-10-31 15:52:22 ✅ no migration for TransactionStorage 2021-10-31 15:52:22 ⚠️ Uniques declares internal migrations (which _might_ execute). On-chain `<wasm:stripped>` vs current storage version `<wasm:stripped>` 2021-10-31 15:52:22 Running migration storage v1 for uniques with storage version wasm:stripped 2021-10-31 15:52:22 Running migration storage v1 for uniques with storage version wasm:stripped was complete 2021-10-31 15:52:22 ✅ no migration for Gilt

run this with `--execution Native`

**Comment by shawntabrizi** [comment]
> Looks overall good to me. To test the migration properly with try-runtime you need to integrate it in cumulus, but I think the logic of the migration is pretty trivial. I am just worried about the size of migration.

I think there are very few classes actually, so this should be fine?

File: frame/uniques/src/migration.rs (hunk 2 above)
Comment (hamidra) [reply frame/uniques/src/migration.rs:37]: What is the best practice to bound the migrations. do we just set an upper limit?!

**Comment by hamidra** [comment]
> > Looks overall good to me. To test the migration properly with try-runtime you need to integrate it in cumulus, but I think the logic of the migration is pretty trivial. I am just worried about the size of migration.
> 
> I think there are very few classes actually, so this should be fine?

currently there are 7 classes. so it will be a light migration.

**Comment by hamidra** [comment]
> Looks overall good to me. To test the migration properly with try-runtime you need to integrate it in cumulus, but I think the logic of the migration is pretty trivial. I am just worried about the size of migration.

Now that try-runtime is added to cumulus I also tried the runtime upgrade on cumulus and it was successful against the current statemine state:

2021-12-06 13:17:07 ✅ no migration for Proxy    
2021-12-06 13:17:07 ✅ no migration for Assets    
2021-12-06 13:17:07 ⚠️ Uniques declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(0)`    
2021-12-06 13:17:07 Running migration storage v1 for uniques with storage version StorageVersion(0)    
2021-12-06 13:17:07 Running migration storage v1 for uniques with storage version StorageVersion(0) was complete    
2021-12-06 13:17:10 proof: 9f06aa394eea5630e07c48ae0c9558cef7398f80688cf08b3f5dffa06d8e56ec0c3e65686b6b969ae836a98cfb64a6c690e48d488020721217d71a031cd2d75e12e6279a0869341eccdb5269c5bad2615450113c4d505f0e7b9012096b41c4eb3aaf947f6ea4290800004c5f0684a022a34dd8bfa2baaf44f172b71004018032e57306806680739e8d0bb9146bfb4ece88127889eccc82c5b3bb4ca18a6f088091c51caec933f99099d9533a10fff9fa19d95981a63b7caf99ad17f14b7d606a80b87a3952cfb2e514fbb79173aa18e5bf21c6a3b2fd27f70d42dde0dacda0ea8780366d0f60b7986a88ffd219874f0ac04eb0344850c3d0c775e93ac27702ce4045785f09cce9c888469bb1a0dceaa129672ef83065092473746174656d696e659f02f224bb127609e7cebd7d411456bd968a6e806d47c287ea48da6148df16f6021539c89aaab786440ff04a34cea13547c7a395805e531b65407ddd65eb11aa12fc4ea7a958040c314a558180143e7f8d8890952e80ecdd9dd5efdab70a1f5946a274ab5dc208e9345ccfdf3efe8ff58f44d47367d280da166af004aa5ce308ca55b9499480f7afa7f72bbdff23d94e9e7cef79f4a14480245647965fd6c1cddfb5e109c6c951421cba9bff690e40850b907f387b1cc2b780491cc11505735d5484f305ec4ef0d3e13d42858060f976b6a918b333a3ffb95380bc0550ea9c...27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31ac69961f64020000000000000000000000610000006100000000000000006623b1a14cb49e4d32c1f7dff3396751450000007902ce10c51e89ab2907983cc8b846d9d1afa2f9700db18e176c7fcfe709d0406339ce10c51e89ab2907983cc8b846d9d1afa2f9700db18e176c7fcfe709d0406339ce10c51e89ab2907983cc8b846d9d1afa2f9700db18e176c7fcfe709d0406339ce10c51e89ab2907983cc8b846d9d1afa2f9700db18e176c7fcfe709d04063396ca55f4aee000000000000000000000000010000000100000000000000006619bf46066d74446ee15c8c2715f7030d0000007902c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b31c6fd5cedb58f27615450c9c7ac04323db766688b2d8112c4aa008c6b63f80b3192b9626480030000000000000000000000aa000000aa0000000000000001 / 36 nodes    
2021-12-06 13:17:10 proof size: 626.45 KB (641488 bytes)    
2021-12-06 13:17:10 compact proof size: 625.35 KB (640360 bytes)    
2021-12-06 13:17:10 zstd-compressed compact proof 625.37 KB (640381 bytes)    
2021-12-06 13:17:10 TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = 2325000000, total weight = 500000000000 (0.00465)

**Review by shawntabrizi — APPROVED** [review]

**Review by gavofyork — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**